File: 프로그래머스/unrated/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
		
		Stack<int[]> stack = new Stack<>();
		
		stack.push(new int[] {0, numbers[0]});
		
		for(int i=1; i<numbers.length; i++) {
			if(stack.isEmpty()) {
				stack.push(new int[] {i, numbers[i]});
			} else {
				while(!stack.isEmpty() && stack.peek()[1] < numbers[i]) {
					answer[stack.pop()[0]] = numbers[i];
				}
				stack.push(new int[] {i, numbers[i]});
			}
		}
		
		while(!stack.isEmpty()) {
			answer[stack.pop()[0]] = -1;
		}
        
        return answer;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [0] * len(numbers)
    answer[-1] = -1
    
    list1 = []
    
    for index, data in enumerate(numbers):
        if not list1:
            list1.append((index, data))
        else:
            while list1 and list1[-1][1] < data:
                i, d = list1.pop()
                answer[i] = data
                
            list1.append((index, data))
        
    
    while list1:
        i, d = list1.pop()
        answer[i] = -1       
    
    return answer